### ISSUE TITLE
When creating a fork, mempool is evaluated wrt to wrong ledger state

### ISSUE THREAD

**Issue by edsko** [input-output-hk/ouroboros-network#1437]
If two leaders are elected for the same slot, we _could_ get the block from the other node before we manage to produce ours. When this happens, we should create a fork, connecting our new block to the _previous_ block, rather than appending it to the block we received. 

We currently do this, but we evaluate the mempool with respect to the wrong ledger state, so that we might produce an invalid block. Instead, we should evaluate it wrt to the previous ledger state.

This has a few consequences:

* When we applied the incoming block, we will likely have removed some transactions from our mempool, and so we will not include them in our own block.
  
  This will be mitigated to some degree by the thread that monitors the chain and reintroduces transactions that get removed by a switch-to-fork (#1269).
  
  However, even with that in place, we might still want to consider doing something special here. As it stands, that thread would notice some transactions got lost _after_ we produce our block, so that they might be included in the _next_. Ideally however we would include them in _this_ block.

  Note that from the perspective of other nodes, the situation is slightly different. They will either adopt the block produced by the other node, in which case anything we decide to do is irrelevant, or they will produce the block that we produce, in which case any transactions that may be missing will probably still exist in their mempool (assuming that all transactions are sent to all core nodes).
  
* A related problem is that the mempool can contain dependent transactions. After applying the incoming block, the mempool might now contain transactions that are valid with respect to the ledger state after applying the incoming block, but will be _invalid_ when evaluated against the previous ledger state (since the transactions they depend on have not been applied in that state). This means that when we evaluate the mempool with respect to that prior ledger state, those dependent transactions from the mempool will be removed as invalid, and not be included in the block we produce.

   **This is not addressed by the thread mentioned in the previous point**: the transactions that are deleted from the mempool are not included in the incoming block, _nor_ in our block, and so we would not reintroduce them into the mempool simply if we monitor the chain.
   
   From the perspective of another node, the situation is again slightly different. As above, if they don't end up adopting our block, what we decide here doesn't really matter. If they do, they simply wouldn't remove _either_ the transactions that had been included in the block produced by the other node, _nor_ the ones that we removed because they depended on them, and so they would be able to include all of them in the next block. This does however depend on other codes having the same transactions in their mempool.

* In order to be to correctly evaluate the mempool in this case, we must be able to evaluate it against the previous ledger state. This means that it's relatively important that we can efficiently get that previous ledger from the chain DB (#1026).

**Comment by edsko** [comment]
This is not strictly speaking high priority for Byron, even though it's a bug, because there will be no multiple slot leaders. However, fixing this will be required for #1194, which is how I discovered this problem in the first place, and that ticket _is_ high priority for Byron.

**Comment by edsko** [comment]
An alternative attempt to fix this problem might be to adopt blocks at the _end_ of the slot instead of at the beginning. This is problematic, however: it is important that we get blocks to the slot leader in time, so that they can connect their block to ours. If we only adopt a block we forge at the end of a slot, this cannot happen. Introducing a special case for blocks we forge, as opposed to incoming blocks, doesn't help, because in a setting where the network is not fully connected, we might sit in the middle of two other nodes

```
A -- us -- C
```

and if `C` is the slot leader in the next slot, it is important that we forward the block we got from `A`  to `C` in time, and not at the end of the slot. Support for tentative headers (#504) could _conceivably_ be used to address this, but now it starts getting messier again. 

And even with all that, it doesn't really fix the whole problem. Consider a Praos setting, with a much shorter slot length. A block comes in for the same slot that we produce one for, and we only adopt it in the next slot (i.e., 2 seconds later). But our system is under such heavy load that we only _start_ producing our block in that next slot (this would still be fine, many slots in Praos are not filled and so if we produce our block a bit too late, it's not necessarily a problem, as long as we get it to the next slot leader before _they_ produce their block). Now the ledger state has already been updated, and we're back to the same problem we started with.
